internal/speakerlist/speakerlist.go

```diff
@@ -80,36 +83,107 @@ func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace
 }
 
 // Start starts the needed goroutines
-func (sl *SpeakerList) Start(client *k8s.Client) error {
+func (sl *SpeakerList) Start(client *k8s.Client) {
 	sl.client = client
 
 	if sl.mList == nil {
-		return nil
+		return
 	}
 
 	go sl.watchEvents()
+	go sl.memberlistJoinLoop()
+}
 
-	iplist, err := sl.client.GetPodsIPs(sl.namespace, sl.labels)
-	if err != nil {
-		sl.l.Log("op", "startup", "error", err, "msg", "failed to get PodsIPs")
-		return err
+func (sl *SpeakerList) memberlistJoin() {
+	if sl.mList == nil {
+		return
+	}
+	joinList := sl.joinList()
+	if len(joinList) > 0 {
+		n, err := sl.mList.Join(joinList)
+		sl.l.Log("op", "ml-join", "msg", "Memberlist join", "nb joigned", n, "error ?", err, "join list", joinList)
+	}
+}
+
+func (sl *SpeakerList) joinList() []string {
+	joinList := []string{}
+	sl.Lock()
+	defer sl.Unlock()
+	if len(sl.k8sSpeakers) == 0 {
+		return joinList
+	}
+
+	mlSpeakers := map[string]bool{}
+	for _, n := range sl.mList.Members() {
+		mlSpeakers[n.Name] = true
+	}
+	for name, ip := range sl.k8sSpeakers {
+		if _, ok := mlSpeakers[name]; !ok {
+			joinList = append(joinList, ip)
+		}
+	}
+	return joinList
+}
+
+func (sl *SpeakerList) memberlistJoinLoop() {
+	ticker := time.NewTicker(10 * time.Second)
+	for {
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			sl.memberlistJoin()
+		}
 	}
-	n, err := sl.mList.Join(iplist)
-	sl.l.Log("op", "startup", "msg", "Memberlist join", "nb joigned", n, "error ?", err)
+}
 
-	return nil
+// SetSpeakersK8S update the list of speakers from the K8S API point of view
+func (sl *SpeakerList) SetSpeakersK8S(eps *v1.Endpoints) {
+	newmap := map[string]string{}
+	for _, subset := range eps.Subsets {
+		for _, ep := range subset.Addresses {
+			newmap[*ep.NodeName] = ep.IP
```

Comment:
No need to skip nodes without name, as done here: https://github.com/metallb/metallb/blob/734ee6743ebe9a8961a95a73ecb24b7f58f0e684/speaker/layer2_controller.go#L54-L56 ?

Reply:
thanks, I'll have a look